### PR TITLE
Use previous Title/Description option added

### LINKED ISSUE #2035: Add "use previous title/desc" option to new upload screen (2.9)
The new upload process (2.9) is missing the "use previous title/desc" option. We would also need to discuss where in the UI it would fit best, and what sort of UI element to use.

### PR DESCRIPTION
**Pull request by nfatma**
Fixes #2035  Add "use previous title/desc" option to new upload screen (2.9)

What changes did you make and why?
1. Added a button in the activity_upload_bottom_card to get the title and description from the previous upload which is visible only if there was at least one attempt of contribution. 
2. Use of BasicKvStore to store the previous title and description in the UploadActivity

Screenshot_20190322-225134

### CODE REVIEW

File: app/src/main/res/layout/activity_upload_bottom_card.xml [hunk 1]
```diff
@@ -132,6 +132,22 @@
...
+            <Button
+                android:id="@+id/prev_title_desc"
+                android:layout_width="match_parent"
+                android:layout_height="29dp"
+                android:layout_marginStart="8dp"
+                android:layout_marginTop="10dp"
+                android:layout_marginBottom="5dp"
+                android:height="10dp"
+                android:background="@color/white"
+                android:text="Copy Title/Desc from previous Upload &amp; Edit "
```
Comment (maskaravivek) [inline app/src/main/res/layout/activity_upload_bottom_card.xml]: Extract to a string resource.

File: app/src/main/java/fr/free/nrw/commons/upload/UploadPresenter.java [hunk 2]
```diff
@@ -115,7 +115,11 @@ void handleNext(Title title,
                     List<Description> descriptions) {
         Timber.e("Inside handleNext");
         view.showProgressDialog();
+        prevTitleDescription
+        setTitleAndDescription(title, descriptions);
+        uploadModel.getImageQuality(uploadModel.getCurrentItem(), true)
         compositeDisposable.add(uploadModel.getImageQuality(uploadModel.getCurrentItem(), true)
+        master
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadPresenter.java]: is this intended?

File: app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java [hunk 3]
```diff
@@ -713,4 +744,14 @@ public void showSimilarImageFragment(String originalFilePath, String possibleFil
         newFragment.setArguments(args);
         newFragment.show(getSupportFragmentManager(), "dialog");
     }
+
+    public void saveOnSubmit(){
+        flag++;
+        BasicKvStore storePrevTitleDesc = new BasicKvStore(this, "prevTitleDesc");
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java]: Use `directKvStore` instead of creating a new store.

File: app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java [hunk 4]
```diff
@@ -713,4 +744,14 @@ public void showSimilarImageFragment(String originalFilePath, String possibleFil
...
+        storePrevTitleDesc.putString("title", descriptionsAdapter.getTitle().toString());
+        for(int i = 0; i < descriptionsAdapter.getItemCount() - 1; i++) {
+            storePrevTitleDesc.putString("description"+Integer.toString(i), descriptionsAdapter.getDescriptions().get(i).getDescriptionText());
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java]: Instead of storing `description0`, `description1` etc, store `description_<language_code>`.

File: app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java [hunk 5]
```diff
@@ -519,6 +532,23 @@ private void configureRightCard() {
...
+    public void configurePrevButton(){
+        prevTitleDecs.setOnClickListener((View v) -> {
+            BasicKvStore storePrevTitleDesc = new BasicKvStore(this, "prevTitleDesc");
+            String title = storePrevTitleDesc.getString("title");
+            Title t = new Title();
+            t.setTitleText(title);
+
+            Description description= new Description();
+            List<Description> finalDesc = new LinkedList<>();
+            String desc = storePrevTitleDesc.getString("description0");
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java]: Set descriptions in all languages that are saved. ie for all `description_<language_code>`.

File: app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java [hunk 6]
```diff
@@ -505,6 +510,14 @@ private void configureTopCard() {
     }
 
     private void configureBottomCard() {
+        BasicKvStore flag = new BasicKvStore(this, "prevTitleDesc");
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java]: use `defaultKvStore` instead of creating new storage.

File: app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java [hunk 7]
```diff
@@ -519,6 +532,23 @@ private void configureRightCard() {
         rightCardMapButton.setOnClickListener(v -> presenter.openCoordinateMap());
     }
 
+    public void configurePrevButton(){
+        prevTitleDecs.setOnClickListener((View v) -> {
+            BasicKvStore storePrevTitleDesc = new BasicKvStore(this, "prevTitleDesc");
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java]: use `defaultKvStore` instead of creating new storage.

File: app/src/main/res/layout/activity_upload_bottom_card.xml [hunk 8]
```diff
@@ -132,6 +132,22 @@
...
+                android:textColor="@color/button_blue"
+                app:layout_constraintBottom_toTopOf="@id/rv_descriptions"
+                app:layout_constraintStart_toStartOf="parent"
+                app:layout_constraintTop_toBottomOf="@id/bottom_card_subtitle"
+                app:layout_constraintVertical_bias="1.0" />
+
```
Comment (neslihanturan) [inline app/src/main/res/layout/activity_upload_bottom_card.xml]: Your screenshot seems problematic. The button is alligned to right.

File: app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java (hunk 7 above)
Comment (nfatma) [reply app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java]: okay I'll implement all your points.

File: app/src/main/res/layout/activity_upload_bottom_card.xml (hunk 8 above)
Comment (nfatma) [reply app/src/main/res/layout/activity_upload_bottom_card.xml]: yes, I'll change it.

**Comment by maskaravivek** [comment]
@nfatma Can you post the updated screenshot/video where copy/pasting multiple descriptions work.

**Comment by nfatma** [comment]
I'll post the video of it.👍

**Comment by nfatma** [comment]
Video of "Use Title/desc from previous upload & edit" button funtioning.

**Comment by maskaravivek** [comment]
@nfatma I encountered this crash while testing your PR. 

```
2019-03-30 14:27:17.988 10597-10597/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: fr.free.nrw.commons.beta, PID: 10597
    java.lang.IllegalArgumentException: Scrapped or attached views may not be recycled. isScrap:false isAttached:true fr.free.nrw.commons.widget.HeightLimitedRecyclerView{a838269 VFED..... ......ID 22,258-1058,566 #7f0901d1 app:id/rv_descriptions}, adapter:fr.free.nrw.commons.upload.DescriptionsAdapter@4651c68, layout:androidx.recyclerview.widget.LinearLayoutManager@e24f881, context:fr.free.nrw.commons.upload.UploadActivity@1d40dba
        at androidx.recyclerview.widget.RecyclerView$Recycler.recycleViewHolderInternal(RecyclerView.java:6267)
        at androidx.recyclerview.widget.RecyclerView$Recycler.recycleView(RecyclerView.java:6203)
        at androidx.recyclerview.widget.RecyclerView$LayoutManager.removeAndRecycleViewAt(RecyclerView.java:8736)
        at androidx.recyclerview.widget.RecyclerView$LayoutManager.removeAndRecycleAllViews(RecyclerView.java:10164)
        at androidx.recyclerview.widget.RecyclerView.removeAndRecycleViews(RecyclerView.java:1171)
        at androidx.recyclerview.widget.RecyclerView.setAdapterInternal(RecyclerView.java:1194)
        at androidx.recyclerview.widget.RecyclerView.setAdapter(RecyclerView.java:1153)
        at fr.free.nrw.commons.upload.UploadActivity.lambda$configurePrevButton$13(UploadActivity.java:549)
        at fr.free.nrw.commons.upload.-$$Lambda$UploadActivity$B7jx74beC4SjeiOAGL4tKO_KU_4.onClick(Unknown Source:2)
        at android.view.View.performClick(View.java:6312)
        at android.view.View$PerformClick.run(View.java:24811)
        at android.os.Handler.handleCallback(Handler.java:790)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:171)
        at android.app.ActivityThread.main(ActivityThread.java:6651)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:547)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:824)

```

It didn't happen for my first 3 tests but happened when I was trying it for the 4th time. This time when I clicked on the copy button, the app crashed.

**Comment by maskaravivek** [comment]
Another issue is that when I copy descriptions in multiple languages, the default language `af` is selected in the dropdown instead of the correct language. 

device-2019-03-30-143056

**Comment by nfatma** [comment]
I understood about the default language being selected, will change it. But I am not sure why the app crashed, will look and notify soon.

**Comment by maskaravivek** [comment]
@nfatma Is it ready for review?

**Comment by nfatma** [comment]
I think I need a couple days to check for the crash that you mentioned.

**Comment by misaochan** [comment]
Thanks for submitting this PR. :) I think the UI needs a bit of work, though. 

The button is far too verbose IMO, we should not be having that much text in one button. I understand that @nicolas-raoul requested that 

> About the wording: It would be great if we could convey the meaning of "reuse the text you entered and modify it to fit this particular picture" rather than just "reuse the text as an exact duplicate".

But I really don't think that putting all the text into the button is the way to go.

I suggest putting the button below the description field, and just having it say "use previous title/desc". We can have an "i" tooltip next to it similar to the ones in title and desc, and when user taps on the "i" tooltip we can say  "Please modify the text to fit this particular picture"?

**Comment by neslihanturan** [comment]
Actually this feature was exists on previous versions, and it was working fine. Since solving this issue gained some priority, maybe searching for the solution from theold version and implementing in that way can help @nfatma

**Comment by nfatma** [comment]
@misaochan thank you for the suggestions :) 
Adding a tooltip to the button is a great idea and I was thinking, It would be nice if users can spot the button right before entering title/description, so may be placing it above is apt otherwise we can completely go as per your suggestion.

@neslihanturan thanks :)
That's a great idea, I searched for the older solution and found that I have done pretty same thing as there. I think I'll have to search a bit more.

**Comment by misaochan** [comment]
Hmm, having the button at the top could work, I guess. :) Feel free to leave it there after the other modifications are done and we can see how it looks. Maybe it just looks odd now because of the extremely long button, haha.

**Comment by misaochan** [comment]
Also, please make sure that this also works for multiple uploads. For instance, if I upload 3 pictures of a cow, I should be able to type "Cow" into the title field of Step 1 for image number 1, then in Step 2 for image number 2 I should be able to tap the "use previous title/desc" button and have it retrieve the title and desc for Cow (image number 1), not for my previous **upload**. However if the user taps that button in Step 1 for image 1, then retrieving the previous upload is OK.

Basically, in multiple uploads the title/desc of the **previous picture in the set** should take precedence over the title/desc of the previous upload.

**Comment by maskaravivek** [comment]
> Basically, in multiple uploads the title/desc of the previous picture in the set should take precedence over the title/desc of the previous upload.

Basically it should simply override. It should save just the last entered title and description. The last can either be from previous upload or previous picture.

**Comment by nfatma** [comment]
@misaochan  @maskaravivek  Roger that
I'll implement them as soon.

**Comment by misaochan** [comment]
Hello @nfatma , would it be possible to get a status update please? :) We intend to include this in v2.10.3 if possible.

**Comment by nfatma** [comment]
Yes, I am currently working on it and I'll be able to push the code for testing by tomorrow.

**Comment by nfatma** [comment]
I have added the functionality for multiple upload as described by @misaochan and @maskaravivek.
Please review it.
UI still need some work, should I add the tooltip?

**Comment by misaochan** [comment]
@nfatma Yes, please add the tooltip as mentioned. :) Ping me when you are done with that, and I will test the multiple upload functionality at the same time.

File: app/src/main/res/values/strings.xml [hunk 9]
```diff
@@ -531,6 +531,7 @@ Upload your first media by tapping on the add button.</string>
   <string name="images_featured_explanation">Featured pictures are images from highly skilled photographers and illustrators that the Wikimedia Commons community has chosen as some of the highest quality on the site.</string>
   <string name="images_via_nearby_explanation">Images Uploaded via Nearby places are the images which are uploaded by discovering places on the map.</string>
   <string name="thanks_received_explanation" >This feature allows editors to send a Thank you notification to users who make useful edits – by using a small thank link on the history page or diff page.</string>
-  <string name="get_previous_title_description">Copy Title/Desc from previous Upload &amp; Edit</string>
+  <string name="previous_upload_title_description">Title/Desc of previous Upload</string>
+  <string name="prprevious_image_title_description">Title/Desc of previous Image</string>
```
Comment (misaochan) [inline app/src/main/res/values/strings.xml]: I don't think there is a need to separate these two. "Use previous title/desc" works fine for both cases IMO.

Comment (misaochan) [reply app/src/main/res/values/strings.xml]: Also, please don't use nondescriptive string names like "prprevious_image_title_description" ;)

Comment (nfatma) [reply app/src/main/res/values/strings.xml]: okay I'll mend them :)

**Comment by misaochan** [comment]
Hi @nfatma , thanks for making the required changes. :) I'm happy to report that it works well for the most part. I tested with an Android 8.0 real device - single uploads, multiple uploads, and uploads via Share. All worked well.

My only concern now is with the UI:

Screenshot_20190415-204639_Commons

- Looking at it now, I think the clearest text for the button should be "Copy previous title & description". Please don't leave the verb out, as it is not clear otherwise.
- I strongly feel the button should be below the title and desc fields, not above. When you put it above them in this way, it looks like a header - not the message we want to send

@maskaravivek and @neslihanturan , can you guys confirm that the problems you reported have been fixed? If you are happy with this PR, I will merge it once my UI requests are resolved. :)

**Comment by nfatma** [comment]
Thank You so much @misaochan for the review:D and regarding the UI I'll change them as per your request:)

**Comment by misaochan** [comment]
No worries. Please post a screenshot when you are done. :)

**Comment by nfatma** [comment]
Screenshots of the button at bottom and 
-centre aligned
Screenshot_20190416-011022

-left aligned
Screenshot_20190416-011231

File: app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java [hunk 10]
```diff
@@ -129,6 +138,7 @@
     private DescriptionsAdapter descriptionsAdapter;
     private RVRendererAdapter<CategoryItem> categoriesAdapter;
     private ProgressDialog progressDialog;
+    private int flag = 0, s_count = 0;
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java]: @nfatma What are these variables for? Can you rename it to be more descriptive?

File: app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java [hunk 11]
```diff
@@ -472,6 +496,49 @@ private void configureRightCard() {
         rightCardMapButton.setOnClickListener(v -> presenter.openCoordinateMap());
     }
 
+    @SuppressLint({"NewApi", "ClickableViewAccessibility"})
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java]: why are we suppressing `NewApi` list warning? It could lead to crashes on lower Android versions. Please check this.

File: app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java [hunk 12]
```diff
@@ -480,6 +547,9 @@ private void configureNavigationButtons() {
                 return;
             }
             setTitleAndDescriptions();
+            if (s_count > 3) {
```
Comment (maskaravivek) [inline app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java]: What is the purpose of `s_count > 3` condition? Lets not use [magic numbers](https://en.wikipedia.org/wiki/Magic_number_(programming)?oldformat=true) in code.

**Comment by misaochan** [comment]
Thanks @nfatma , I prefer center-aligned. :) I'll go ahead and approve your PR now, and then I'll merge it once Vivek's issues are resolved.

File: app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java (hunk 10 above)
Comment (nfatma) [reply app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java]: yes, alright:)

File: app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java (hunk 12 above)
Comment (nfatma) [reply app/src/main/java/fr/free/nrw/commons/upload/UploadActivity.java]: s_count > 3 is used to check the condition of multiple uploads. I'll try to figure out an alternative.

**Comment by misaochan** [comment]
Tested and works well. Thanks @nfatma ! Congrats on your first contribution. :)
